### ISSUE TITLE
String > Float > String amount conversion breaks decimal place check

### ISSUE THREAD

**Issue by TedAvery** [thephpleague/omnipay-common#69]
I have an amount of $79.10 that I am trying to charge via [omnipay-stripe](https://github.com/thephpleague/omnipay-stripe).

There are a number of actions in [AbstractRequest.php](https://github.com/thephpleague/omnipay-common/blob/5fa50e04275d354bceb0870c0bf1cb6421b3a6f7/src/Omnipay/Common/Message/AbstractRequest.php#L325) starting at line 325 that cause this amount to fail.

If I pass the string `'79.10'` as an amount when creating the charge, line 325 converts it to float, which causes it to become a float of `79.09999999999999`. But then at line 338, it takes that amount that was converted to float and puts it back to string (`'79.09999999999999'`) which then fails the decimal place check, throwing the exception `'Amount precision is too high for currency.'`

So as it stands, $79.10 is an impossible amount to charge with this library.

**Comment by judgej** [comment]
That same value is dealt with here:

http://stackoverflow.com/questions/34325753/convert-string-to-float-without-losing-precision

The solution here is going to be about how the float gets rounded in the conversion back to a string.

Floats are a really bad way to store and perform arithmetic on currency values. However, in OmniPay there is no arithmetic, and it does not get stored, so we should be able to cope with this inaccuracy as it should not (in theory) be accumulative.

**Comment by judgej** [comment]
Your precision is set higher than the normal PHP default, so this inaccuracy in storing "0.1" as binary is showing up.

~~~php
// Normal precision.
echo (float)"79.10";

ini_set('precision', 17);

// Highest precision.
echo (float)"79.10";

// Back to a string as is now.
echo (string)((float)"79.10");

// Using number_format to round and then trim the trailing zeros.
echo rtrim(number_format((float)"79.10", 10, '.', ''), '0');
~~~

> 79.1
> 79.099999999999994
> 79.099999999999994
> 79.1

The zero-trimming seems a little ugly. We get 79.1000000000 if we don't do that. I wonder if `sprintf()` will be a better option?

Moving to a currency library that uses integers to handle currency, for OmniPay 3.0, will be a welcome move.

Your quickest workaround if you need one immediately, will be to wind the precision down a tad. For me, even a precision of 15 gets it working again.

**Comment by judgej** [comment]
~~~php
echo sprintf('%.10g', (float)"79.10");
// 79.1
~~~

The precision is an arbitrary 10 digits, so we do catch digits way off the currency's precision (minor units), but is far enough away from the precision errors that PHP float/double has when storing simple decimals ending with a "1" digit. Maybe 8 digits would be a better compromise?

I can push a fix for this, probably using sprintf, and we need a test or two that wind the precision up to 17 digits. Any thoughts, bearing in mind this is a fix for what we have in 2.0 now, and not a refactor to get rid of the float entirely?

**Comment by TedAvery** [comment]
Thanks @judgej, that StackOverflow question is actually mine ;) Because of the restrictions in this library converting from String > Float > String I needed an actual float value in the end that would hold the precise 79.10 amount. Your suggestion of `ini_set('precision', 15);` seems to be the fix in my case. But it would be great if the library itself was set up to catch something like this.

**Comment by judgej** [comment]
I should have spotted that -  your mugshot was there and everything :-)

Yes, a reliable fix is needed, and I'll put that on my list for a quiet moment over Christmas.

**Comment by judgej** [comment]
I've used an arbitrary precision of 8dp when converting from `float` to `string`. It just needed to be something higher than any real currency, and high enough to catch any really duff values (such as 0.5 cents) without it being rounded up or down invisibly.

**Comment by judgej** [comment]
@TedAvery let me know what you think of PR #70

**Comment by barryvdh** [comment]
Closed by PR #70
